Applications/Spire/Include/Spire/OrderImbalanceIndicator/OrderImbalanceIndicatorContextMenu.hpp

```diff
@@ -0,0 +1,87 @@
+  class OrderImbalanceIndicatorContextMenu : public QMenu {
+    public:
+
+      //! Signals that a menu option was selected.
+      using SelectedSignal = Signal<void ()>;
+
+      //! Signals that a column check box was toggled.
+      /*
+        \param is_checked True if the check box is in the checked state.
+      */
+      using ToggledSignal = Signal<void (bool is_checked)>;
+
+      //! Constructs an OrderImbalanceINdicatorContextMenu.
+      /*
+        \param parent The parent widget.
+      */
+      OrderImbalanceIndicatorContextMenu(QWidget* parent = nullptr);
+
+      //! Connects a slot to the export table signal.
+      boost::signals2::connection connect_export_table_signal(
+        const SelectedSignal::slot_type& slot) const;
+
+      //! Connects a slot to the export chart signal.
+      boost::signals2::connection connect_export_chart_signal(
+        const SelectedSignal::slot_type& slot) const;
+
+      //! Connects a slot to the reset filters signals.
+      boost::signals2::connection connect_reset_signal(
+        const SelectedSignal::slot_type& slot) const;
+
+      //! Connects a slot to the security toggle signal.
+      boost::signals2::connection connect_security_toggled_signal(
+        const ToggledSignal::slot_type& slot) const;
+
+      //! Connects a slot to the side toggle signal.
+      boost::signals2::connection connect_side_toggled_signal(
+        const ToggledSignal::slot_type& slot) const;
+
+      //! Connects a slot to the size toggle signal.
+      boost::signals2::connection connect_size_toggled_signal(
+        const ToggledSignal::slot_type& slot) const;
+
+      //! Connects a slot to the side reference price toggle signal.
+      boost::signals2::connection connect_reference_price_toggled_signal(
+        const ToggledSignal::slot_type& slot) const;
+
+      //! Connects a slot to the date toggle signal.
+      boost::signals2::connection connect_date_toggled_signal(
+        const ToggledSignal::slot_type& slot) const;
+
+      //! Connects a slot to the time toggle signal.
+      boost::signals2::connection connect_time_toggled_signal(
+        const ToggledSignal::slot_type& slot) const;
+
+      //! Connects a slot to the notional value toggle signal.
+      boost::signals2::connection connect_notional_value_toggled_signal(
+        const ToggledSignal::slot_type& slot) const;
+
+    protected:
+      bool eventFilter(QObject* watched, QEvent* event) override;
+
+    private:
+      mutable SelectedSignal m_export_table_signal;
+      mutable SelectedSignal m_export_chart_signal;
+      mutable SelectedSignal m_reset_signal;
+      mutable ToggledSignal m_security_signal;
+      mutable ToggledSignal m_side_signal;
+      mutable ToggledSignal m_size_signal;
+      mutable ToggledSignal m_ref_px_signal;
+      mutable ToggledSignal m_date_signal;
+      mutable ToggledSignal m_time_signal;
+      mutable ToggledSignal m_notional_value_signal;
+      QMenu* m_table_columns_menu;
```

Comment:
Good point in general, unfortunately Qt uses a very old style of C++ programmer and so we have to balance modern C++ coding style with old pre-C++11 coding styles.

Qt has its own object ownership model that it uses for objects that inherit from ```QObject``` so for ```QObject```s we use Qt's ownership model. That model does not make use of ```unique_ptr``` but rather every ```QObject``` has a ```parent``` it will attach itself to, which is why many ```QObject``` constructors take in a ```parent``` parameter that is defaulted to ```nullptr```.

Reply:
OK, but it does not seem like we cannot use `std::unique_ptr` along with `get()` and `release()` methods to adapt to Qt style where composition is concerned. So it just a convention for Spire?

Reply:
It ends up being awkward using two orthogonal ownership models for ```QObjects``` that we have to keep track of, and so for ```QObjects``` we use Qt's ownership model and for everything else we use more conventional RAII/smart pointers etc... Unfortunately you'll find many other instances of this, where for Qt there is one set of rules, and for everything else there is another set of rules. It's just an artifact of the fact that Qt is a very old framework that hasn't kept up that well with modern C++.

Reply:
Also, regardless of the convention, `m_table_columns_menu` is only passed to `addMenu`, which does not take ownership. Where will `m_table_columns_menu` be destroyed?

Reply:
```m_table_columns_menu``` is allocated on line 60. As per Qt's object ownership model, it is owned by ```this``` so it will get destroyed when ```this``` gets destroyed. That's why all ```QObject```s have that ```parent``` parameter that is defaulted to ```nullptr```, because ```QObject```'s form an ownership hierarchy. I agree in principle that it's frustrating that Qt works this way given how modern C++ codebases establish object ownership, but this is one of those areas where it's better to use Qt's way of doing things when dealing with ```QObject```s rather than mixing two different ways that end clashing with one another.

Reply:
Got it. The same applies to all those local `QWidgetAction`'s, right?

Reply:
Yes exactly, the ```QWidgetAction```s are all owned by ```this```. All the points you made are good valid points under normal circumstances, as we do more code reviews we will just have to converge onto a uniform coding style for UI code and code involving Qt and then we'll update our coding style on the wiki. Admittedly this will be frustrating for you but I don't want to discourage you from making recommendations for good coding style, on the other hand UI code is intrinsically a lot messier and free form. So the first few code reviews involving UI code will be a work in progress so to speak as we settle on a coding convention for UI applications.

Reply:
I just overlooked that `this` magic and made wrong assumptions about what Qt will be capable of doing in terms of memory management. Sorry for that.